### PR TITLE
added assertions for InputStream: hasSameContentAs, hasNotSameContentAs

### PR DESCRIPTION
**Pull request by dibog**
contains hasSameContentAs and hasNotSameContentAs assertions for InputStreams

### CODE REVIEW

**Comment by dibog** [comment]
added support for Path

**Comment by dibog** [comment]
all tests passed in CircleCI

File: src/main/kotlin/assertk/assertions/inputstream.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import assertk.assertions.support.expected
+import java.io.InputStream
+
+fun Assert<InputStream>.hasSameContentAs(other: InputStream) {
+    val buf1 = ByteArray(4096)
+    val buf2 = ByteArray(4096)
+
+    var size = 0L
+    while(true) {
+        val len1 = actual.read(buf1)
```
Comment (evant) [inline src/main/kotlin/assertk/assertions/inputstream.kt]: Don't think this test is right. `read` doesn't define how many bytes it will read each time https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[]). Check out the implementation in apache commons http://commons.apache.org/proper/commons-io/javadocs/api-release/src-html/org/apache/commons/io/IOUtils.html#line.2551

File: src/main/kotlin/assertk/assertions/inputstream.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        val len2 = other.read(buf2)
+
+        when {
+            len1<len2 -> expected("to have the same size, but actual has size ${size+len1} which is smaller then the expected stream")
```
Comment (evant) [inline src/main/kotlin/assertk/assertions/inputstream.kt]: Think it would be more useful to consume the rest of the other stream and report both sizes. ex:
`expected("to have the same size, but expected was:${expectedSize} and actual was:${actualSize}")

File: src/main/kotlin/assertk/assertions/inputstream.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            len1>len2 -> expected("to have the same size, but the expected stream size ${size+len2} which is larger then the acual stream")
+            len1==-1 -> return
+        }
+
+        for(i in 0..len2) {
+            val b1 = buf1[i]
+            val b2 = buf2[i]
+            if(b1!=b2) {
+                expected("to have the same content, but found first difference at pos ${size+i}")
```
Comment (evant) [inline src/main/kotlin/assertk/assertions/inputstream.kt]: Should report what the difference is as well as it's position.

File: src/main/kotlin/assertk/assertions/inputstream.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        val len1 = actual.read(buf1)
+        val len2 = other.read(buf2)
+
+        when {
+            len1<len2 -> expected("to have the same size, but actual has size ${size+len1} which is smaller then the expected stream")
+            len1>len2 -> expected("to have the same size, but the expected stream size ${size+len2} which is larger then the acual stream")
+            len1==-1 -> return
+        }
+
+        for(i in 0..len2) {
```
Comment (evant) [inline src/main/kotlin/assertk/assertions/inputstream.kt]: Is this right? Isn't `..` an inclusive range?

File: src/main/kotlin/assertk/assertions/inputstream.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+}
+
+fun Assert<InputStream>.hasNotSameContentAs(other: InputStream) {
+    val buf1 = ByteArray(4096)
+    val buf2 = ByteArray(4096)
+
+    var equals: Boolean
+    while(true) {
+        val len1 = actual.read(buf1)
```
Comment (evant) [inline src/main/kotlin/assertk/assertions/inputstream.kt]: Same issue with `read`

File: src/main/kotlin/assertk/assertions/path.kt [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package assertk.assertions
+
+import assertk.Assert
+import assertk.assertions.support.expected
+import assertk.assertions.support.show
+import java.nio.file.Files
+import java.nio.file.LinkOption
+import java.nio.file.Path
+
+fun Assert<Path>.isRegularFile(vararg options: LinkOption) {
```
Comment (evant) [inline src/main/kotlin/assertk/assertions/path.kt]: Can you put the path assertions in a separate pr? That way one doesn't block the other.

File: src/main/kotlin/assertk/assertions/inputstream.kt [hunk 7]
```diff
@@ -0,0 +1,62 @@
+package assertk.assertions
+
+import assertk.Assert
+import assertk.assertions.support.expected
+import java.io.InputStream
+
+fun Assert<InputStream>.hasSameContentAs(other: InputStream) {
```
Comment (evant) [inline src/main/kotlin/assertk/assertions/inputstream.kt]: I don't know if it was there when you first started looking, but I added docs to all of the assertions. Can you do the same? ex: https://github.com/willowtreeapps/assertk/blob/master/src/main/kotlin/assertk/assertions/boolean.kt

File: src/test/kotlin/test/assertk/assertions/InputStreamSpec.kt [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    given("an empty stream") {
+
+        it("has the same content as another empty stream") {
+            assert(emptyStream()).hasSameContentAs(emptyStream())
+        }
+
+        it("has the same content as another empty stream (using hasNotSameContentAs)") {
+            assertThatThrownBy {
+                assert( emptyStream() ).hasNotSameContentAs( emptyStream() )
```
Comment (evant) [inline src/test/kotlin/test/assertk/assertions/InputStreamSpec.kt]: Can you fix your formatting? The current style doesn't have spaces around fun args. If you are using intellij the format code option should fix things.

File: src/test/kotlin/test/assertk/assertions/InputStreamSpec.kt [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            val copyArray = byteArray.copyOf()
+            val otherStream = ByteArrayInputStream(copyArray)
+
+            assertThatThrownBy {
+                assert( stream() ).hasNotSameContentAs( otherStream )
+            }.hasMessage("expected stream not to be equal, but they were equal")
+        }
+
+        it("has not the same content as another stream using another byte array") {
+            val otherStream = ByteArrayInputStream(randomByteArray())
```
Comment (evant) [inline src/test/kotlin/test/assertk/assertions/InputStreamSpec.kt]: Don't think you need a random byte array. You should test 2 byte arrays with known differences, like actual shorter, expected shorter, and same size but some content different.

File: src/test/kotlin/test/assertk/assertions/PathSpec.kt [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            assert( regularFile ).isRegularFile()
+        }
+
+        it("is not a folder") {
+            assertThatThrownBy {
+                assert( regularFile ).isDirectory()
+            }.hasMessage("expected <LICENSE> to be a directory, but it is not")
+        }
+
+//        it("is not an executable") {
```
Comment (evant) [inline src/test/kotlin/test/assertk/assertions/PathSpec.kt]: Why is this commented out?

File: src/test/kotlin/test/assertk/assertions/PathSpec.kt [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+import org.assertj.core.api.Assertions.assertThatThrownBy
+import org.jetbrains.spek.api.Spek
+import org.jetbrains.spek.api.dsl.given
+import org.jetbrains.spek.api.dsl.it
+import java.nio.file.Paths
+
+class PathSpec: Spek({
+
+    given("a regular file") {
+        val regularFile = Paths.get("LICENSE")
```
Comment (evant) [inline src/test/kotlin/test/assertk/assertions/PathSpec.kt]: Instead of using project files, create temp files for the test.

File: src/main/kotlin/assertk/assertions/path.kt (hunk 6 above)
Comment (evant) [reply src/main/kotlin/assertk/assertions/path.kt]: Also, path isn't in android until api 26, we might want a separate java8 artifact.

File: src/main/kotlin/assertk/assertions/inputstream.kt (hunk 7 above)
Comment (dibog) [reply src/main/kotlin/assertk/assertions/inputstream.kt]: Yep, when I forked it, I couldn't find them. I will add them.

File: src/main/kotlin/assertk/assertions/inputstream.kt (hunk 3 above)
Comment (dibog) [reply src/main/kotlin/assertk/assertions/inputstream.kt]: Imagine a file with a lots of differences. I'm not sure that the information what is different is really useful. I could ofc show the values of the first different byte.

File: src/main/kotlin/assertk/assertions/path.kt (hunk 6 above)
Comment (dibog) [reply src/main/kotlin/assertk/assertions/path.kt]: Regarding separate PRs: How do I do that now?
Regarding separate java8 artiffact: I have even less clue how to do that.

File: src/test/kotlin/test/assertk/assertions/InputStreamSpec.kt (hunk 9 above)
Comment (dibog) [reply src/test/kotlin/test/assertk/assertions/InputStreamSpec.kt]: k, will do that.

File: src/test/kotlin/test/assertk/assertions/PathSpec.kt (hunk 10 above)
Comment (dibog) [reply src/test/kotlin/test/assertk/assertions/PathSpec.kt]: Because under windows files or folders are "executable" (according to Files.isEecutable()) so this test doesn't work like you would expect it

File: src/main/kotlin/assertk/assertions/inputstream.kt (hunk 3 above)
Comment (evant) [reply src/main/kotlin/assertk/assertions/inputstream.kt]: That's what I was thinking, just the first difference, isn't that what it's doing currently?

File: src/main/kotlin/assertk/assertions/path.kt (hunk 6 above)
Comment (evant) [reply src/main/kotlin/assertk/assertions/path.kt]: Don't worry about the java8 artifact, still thinking about the right solution there. For a separate pr, you can branch off this one and delete the inputstream files, then delete the path files from here.

File: src/test/kotlin/test/assertk/assertions/PathSpec.kt (hunk 10 above)
Comment (evant) [reply src/test/kotlin/test/assertk/assertions/PathSpec.kt]: Oh weird, yeah if this method is platform specific it might be hard to test.
